### ISSUE TITLE
Wifi (ftp, http) and ssh connections refused

### ISSUE THREAD

**Issue by colibris79** [spruceUI/spruceOS#256]
Hi!
Just got an A30 and after updating to last firmware I installed SpruceOs and I love what it does with the device (I like it even more than my beloved Miyoo mini).
There is only one issue that I've been trying to figure out but after a few hours I don't know what's going on. 
I have enabled "Wifi file transfer" and SSH in advanced options. The first time I connected I had no issues, but after that I always get a "Connection refused" error and just cannot connect anymore. I've tried connecting through ftp and html, also with my browser. For SSH I've tried both terminal and the file explorer and later deleted the ssh root user and tried again (refused connection). Tried giving an static IP to the A30 but same end result. 
I use Mint 21.3 (Ubuntu 22.04) and I have many devices (including several retro handhelds) connecting without any issues to my PC through ftp and ssh and all of them are working fine. 
And now I cannot access to the root of the sd card since SSH is not working.
Is there anything I can do to solve this issue? (being able to connect through wifi is really handy)
Thanks for all the hard work!

**Comment by veckia9x** [comment]
Sometimes the device IP changes, can you check the IP and try again? Thanks!

**Comment by colibris79** [comment]
Yes. I always check the IP before trying the connection. Still same result. 
And now for some reason I cannot access Advanced settings anymore. When I try to open it just closes. Will try a fresh install with a different SD card later today

**Comment by Sundownersport** [comment]
This sounds like corruption yes. Reformat, reinstall and try again

**Comment by veckia9x** [comment]
Sorry for the delayed response, Sun is right, that sounds like a corrupt sd card.

Do you mind sharing the brand/model and size of the sd card? Thanks!

**Comment by colibris79** [comment]
> Sorry for the delayed response, Sun is right, that sounds like a corrupt sd card.
> 
> Do you mind sharing the brand/model and size of the sd card? Thanks!

No worries. Thanks for the fast response! 
Sandisk 32G. But it wasn't new (though it's the first time a SD card fails like that).

**Comment by colibris79** [comment]
I tried with a different card. And it worked fine for a while, I copied a few roms, rebooted and it still worked. But after copying the rest of my roms I'm facing the same problem. 
I've tried setting a fixed IP in my router but didn't work either. Connection refused.

**Comment by colibris79** [comment]
Ok. I think I am getting somewhere. 
-About the Wifi share: if when I turn on the a30 the wifi is off and I manually turn on the wifi through settings. I'll get the refused connection
If when I turn on the a30 the wifi is on (because it was when I turned off last time), wifi file transfer will work

**Comment by veckia9x** [comment]
Thanks for the info on the SD card. 

> -About the Wifi share: if when I turn on the a30 the wifi is off and I manually turn on the wifi through settings. I'll get the refused connection
> If when I turn on the a30 the wifi is on (because it was on already when I turned the device off last time), wifi file transfer will work

I'll ask the guys to take a look at this case scenario and see if something else might be happening, but I'm pretty sure this was tested and this might be something else still being off place.

> -About last time not being able to open Advanced settings. I like messing with my retroarch configuration because I have it in some many devices that I like to leave it the way I'm so used to. So something in my retroarch.cfg was messing with the advanced settings app (my guess, setting the hotkey to the home button, but will do more testing). Restoring the copy of the original retroarch.cfg solves that problem

I'm not a retroarch advanced guy, but i fail to see how messing with buttons inside retroarch cfg being disruptive on the advanced settings app. I'm glad you found a temporary solution. 

> And, just as a suggestion, may I  recommend  making the turning off screen (those 3 seconds just before the device turns off) have a bit more of contrast or not so dark? I didn't realize till today there was such a screen and I think I may have taken the sd card out while it wasn't completely turned off. Just a suggestion for distracted people like myself : )

I'll ask but im pretty sure this might be just the way A30 works, kill everything app before shutdown, including our code. Just in case I'll spin that suggestion to others. 

Thanks

**Comment by colibris79** [comment]
> 
> I'll ask the guys to take a look at this case scenario and see if something else might be happening, but I'm pretty sure this was tested and this might be something else still being off place.
> 
Cool, thanks. It is happening to me with 2 different sd cards. So if you want me to test anything else, just let me know. But I can confirm that in my case this on/off wifi at startup triggers this issue.

> I'm not a retroarch advanced guy, but i fail to see how messing with buttons inside retroarch cfg being disruptive on the advanced settings app. I'm glad you found a temporary solution.
> 

I can confirm that changing back the hotkey to "Select" inside of retroarch allows me to open "Advanced settings" again. I send you my customised retroarch.cfg in case it helps.

**Comment by XanXic** [comment]
Ah, yeah, 3.1.0 Advanced Settings, is incompatible with Home as a hotkey, I'm curious how you're getting away with using it without firing the tap/hold actions you have defined at the same time? I believe you would've also had to define it outside of Retroarch for that same reason. But there was some work done previously so we can enable that eventually. I didn't anticipate someone setting that up since it's so incompatible with spruce at large before some fixes I made in 3.2.0.

I guess just FYI that work is done in 3.2.0 so home was added as a hotkey in settings and using it as a hotkey will prevent tap and hold settings from firing. But some of the code is in advanced settings 3.1.0 for that lol. So if you set it to literally anything other than home Advanced Settings will be fine.

**Comment by colibris79** [comment]
@XanXic Sorry for all the fuss and thanks for the help. I totally understand that Spruce is designed to make it as easy as possible for most people while having some advanced configurations at the same time. Right now I've got it configured to L2 which is close enough

**Comment by colibris79** [comment]
Just tried a fresh install with new 3.2.0 and still experiencing the same issue
Everything else seems to be working fine (not too much testing yet though). Good that you added the warning bout "Rewind" option. I had to turn it off for psx games a couple of days ago because it made things very choppy.

**Comment by BaseInfinity** [comment]
Out of curiousity what do these files look like for you

![Image](https://github.com/user-attachments/assets/7b656810-a0f3-4db1-8d18-9862ccb8a109)

**Comment by colibris79** [comment]
> Out of curiousity what do these files look like for you
 I do not have a /config/sytem.json. I do have a /spruce/settings/system.json. This is it:
```
{
	"vol":	10,
	"keymap":	"L,L2,R,R2,X,A,B,Y",
	"mute":	0,
	"bgmvol":	0,
	"brightness":	7,
	"language":	"en.lang",
	"hibernate":	0,
	"lumination":	6,
	"hue":	10,
	"saturation":	10,
	"contrast":	10,
	"theme":	"/mnt/SDCARD/Themes/SPRUCE/",
	"fontsize":	24,
	"audiofix":	1,
	"wifi":	1,
	"cpufreq":	0
}
```
And this is the  /SDCARD/.system.json.log
```
{
	"vol": 16,
	"keymap":	"L,L2,R,R2,X,A,B,Y",
	"mute":	1,
	"bgmvol":	0,
	"brightness": 10,
	"language":	"en.lang",
	"hibernate":	0,
	"lumination":	6,
	"hue":	10,
	"saturation":	10,
	"contrast":	10,
	"theme":	"/mnt/SDCARD/Themes/SPRUCE/",
	"fontsize":	24,
	"audiofix":	1,
	"wifi":	0,
	"cpufreq":	0
}
```

"vol" "mute" and "wifi" values are different for some reason

**Comment by BaseInfinity** [comment]
I think the important part here is "wifi": 0, change that to "wifi": 1 in both places and see if that fixes it. Maybe one of the other team members who know more about this could explain why they are out of sync?

**Comment by colibris79** [comment]
> I think the important part here is "wifi": 0, change that to "wifi": 1 in both places and see if that fixes it. Maybe one of the other team members who know more about this could explain why they are out of sync?

Thanks for the suggestion. But the problem remains the same. Changing it to wifi:1 just makes the device start up with the wifi on, which allows the wifi share feature to work. But l have 99% of the time the wifi off, so I don't want to start with the wifi on.
So, just like before, when I start the device with the wifi off and then connect to the wifi, I will still get the "refused connection" error.

**Comment by colibris79** [comment]
Ok. Some progress here. Managed to make it work after restart and with the wifi off doing this:
Turn on the wifi, wait till it connects and I get the IP (wait for a bit)
In terminal:
$ftp
ftp> ftp A30_IP
username: spruce
password: happygaming
230 Password ok, continue
Remote system type is UNIX.
Using binary mode to transfer files.

And after connection successfully I can open the FTP address in both my file or internet browser
